### PR TITLE
Fresh transactionToRequest

### LINKED ISSUE #1313: Make `transactionsToRequest` always fresh
### Description
When a node fills its `transactionsToRequest` queue (at the time of writing the queue size is hardcoded to 10K) new transactions coming in do not get included into the queue and as a result they are never requested.
`transactionsToRequest` should be implemented as a `CircularFifoQueue` of fixed capacity, where old stale transactions are dropped in favor of new inbound ones.

The above may happen:

*  when a node lags considerably behind in syncing, accumulates a lot of stale transactions and denies fresh ones to get in;
* when a lot of floating transactions fill up the queue with stale transactions.

May relate to #779 

### Motivation
Make stale nodes unstuck and / or sync faster.

### Requirements
* This change should be abstracted from all the code but the `TransactionRequester` class.

### Am I planning to do it myself with a PR?
Yes

### PR DESCRIPTION
**Pull request by karimodm**
# Description
Keep the `transactionToRequest` Set of the `TransactionRequester` fresh with new inbound transactions instead of keeping stale ones when at capacity.

Fixes #1313 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

# How Has This Been Tested?

- Unsynced node that lag greatly behind seem to sync faster since we do not pile up stale transations anymore. 

# Checklist:

- [x] My code follows the style guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by GalRogozinski** [comment]
@karimodm 
I approved but maybe add a unit test that checks this very specific change
If you feel like adding other unit tests that don't concern your change do it in another PR so we don't delay the merger of this one

**Comment by karimodm** [comment]
@GalRogozinski wrote tests.

File: src/test/java/com/iota/iri/network/TransactionRequesterTest.java [hunk 1]
```diff
@@ -76,6 +79,48 @@ public void instance() throws Exception {
...
+
+        txReq.popEldestTransactionToRequest();
+        // Check that the transaction is there no more
+        assertEquals(false, txReq.isTransactionRequested(eldest, false));
+    }
+
+    @Test
+    public void transactionRequestedFreshness() throws Exception {
+        // Add some Txs to the pool and see if the method pops the eldest one
+        ArrayList<Hash> eldest = new ArrayList<Hash>(Arrays.asList(
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/network/TransactionRequesterTest.java]:
```suggestion
        List<Hash> eldest = new ArrayList<Hash>(Arrays.asList(
```

Usually in java we like to use the interface for abstraction.
For this unit test it doesn't really matter too much

File: src/test/java/com/iota/iri/network/TransactionRequesterTest.java [hunk 2]
```diff
@@ -76,6 +79,48 @@ public void instance() throws Exception {
...
+                TransactionViewModelTest.getRandomTransactionHash(),
+                TransactionViewModelTest.getRandomTransactionHash(),
+                TransactionViewModelTest.getRandomTransactionHash()
+        ));
+        TransactionRequester txReq = new TransactionRequester(tangle, snapshotProvider, mq);
+        int capacity = TransactionRequester.MAX_TX_REQ_QUEUE_SIZE;
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/network/TransactionRequesterTest.java]: There is a thumb rule that we should declare a local variable as soon as it is being used.
Not going to insist on it now, just that you know for next time.

Change it only if you feel like it

File: src/test/java/com/iota/iri/network/TransactionRequesterTest.java [hunk 3]
```diff
@@ -76,6 +79,48 @@ public void instance() throws Exception {
...
 
+    @Test
+    public void popEldestTransactionToRequest() throws Exception {
+        TransactionRequester txReq = new TransactionRequester(tangle, snapshotProvider, mq);
+        // Add some Txs to the pool and see if the method pops the eldest one
+        Hash eldest = TransactionViewModelTest.getRandomTransactionHash();
+        txReq.requestTransaction(eldest, false);
+        txReq.requestTransaction(TransactionViewModelTest.getRandomTransactionHash(), false);
+        txReq.requestTransaction(TransactionViewModelTest.getRandomTransactionHash(), false);
+        txReq.requestTransaction(TransactionViewModelTest.getRandomTransactionHash(), false);
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/network/TransactionRequesterTest.java]: Just heads up

Technically it is not a good idea to use randomness in unit tests (you want them to be consistent)
Unfortunately, the current codebase encourages you to write this way.

For this specific unit test it is less of a problem and I don't want to currently fix this problem which is not related to you or the task at hand.

Just FYI

Comment (karimodm) [reply src/test/java/com/iota/iri/network/TransactionRequesterTest.java]: I thought the same when I was coding but well...

File: src/test/java/com/iota/iri/network/TransactionRequesterTest.java [hunk 4]
```diff
@@ -76,6 +80,48 @@ public void instance() throws Exception {
...
+        for (int i = 0; i < capacity; i++) {
+            Hash hash = TransactionViewModelTest.getRandomTransactionHash();
+            txReq.requestTransaction(hash,false);
+        }
+
+        //check that limit wasn't breached
+        assertEquals(capacity, txReq.numberOfTransactionsToRequest());
+        // None of the eldest transactions should be in the pool
+        for (int i = 0; i < 3; i++) {
+            assertFalse(txReq.isTransactionRequested(eldest.get(i), false));
```
[the remark this answers is not in the document]
Comment (GalRogozinski) [reply src/test/java/com/iota/iri/network/TransactionRequesterTest.java]: Just this little thing can you fix?
Besides that, all looks good to me :-)

File: src/test/java/com/iota/iri/network/TransactionRequesterTest.java (hunk 2 above)
Comment (GalRogozinski) [reply src/test/java/com/iota/iri/network/TransactionRequesterTest.java]: Since in java we have a heavily optimized compiler readability comes first
For next time :-)

File: src/test/java/com/iota/iri/network/TransactionRequesterTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (GalRogozinski) [reply src/test/java/com/iota/iri/network/TransactionRequesterTest.java]:
```suggestion
            assertFalse("Old transaction has been requested", txReq.isTransactionRequested(eldest.get(i), false));
```

File: src/test/java/com/iota/iri/network/TransactionRequesterTest.java [hunk 5]
```diff
@@ -76,6 +80,48 @@ public void instance() throws Exception {
...
+        for (int i = 0; i < 3; i++) {
+            txReq.requestTransaction(eldest.get(i), false);
+        }
+        for (int i = 0; i < capacity; i++) {
+            Hash hash = TransactionViewModelTest.getRandomTransactionHash();
+            txReq.requestTransaction(hash,false);
+        }
+
+        //check that limit wasn't breached
+        assertEquals(capacity, txReq.numberOfTransactionsToRequest());
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/network/TransactionRequesterTest.java]:
```suggestion
        assertEquals("Queue capacity breached!!", capacity, txReq.numberOfTransactionsToRequest());
```
